hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/batching/PartitionAwareCallable.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.spi.impl.operationservice.impl.batching;
+
+/**
+ * Partition callable that is guaranteed to run on a partition-thread responsible for the given partitionId
```

Comment:
I'm not that crazy about this because it causes a coupling to the concrete threading model and that is exactly what we want to move away from. In the future any thread can run a task on a partition.

Reply:
Yes, I agree. This is sth that should be changed when we introduce an improvement to the threading model. Currently that invokeOnAllPartitions causes a huge stall for quick partition operations. We can agree that this mechanism won't be overused. Now it's the HD-query index, I was also thinking about map.size() operation that is currently 2 times slower than the HD-query index op.

Reply:
I think the contract can be worded along these lines: "The callable is guranteed to have an exclusive access to partition data." 

this way to contract is not bound to a specific threading model yet it offers what the implementer needs.

Reply:
Fixed.